sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/SdkCustomExemplarReservoirTest.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics;
+
+import static io.opentelemetry.api.common.AttributeKey.stringKey;
+import static io.opentelemetry.sdk.testing.assertj.OpenTelemetryAssertions.assertThat;
+
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.metrics.DoubleCounter;
+import io.opentelemetry.api.metrics.Meter;
+import io.opentelemetry.sdk.metrics.internal.exemplar.ExemplarFilter;
+import io.opentelemetry.sdk.metrics.internal.exemplar.ExemplarReservoirFactory;
+import io.opentelemetry.sdk.metrics.internal.view.AggregationExtension;
+import io.opentelemetry.sdk.resources.Resource;
+import io.opentelemetry.sdk.testing.exporter.InMemoryMetricReader;
+import io.opentelemetry.sdk.testing.time.TestClock;
+import java.util.Random;
+import org.junit.jupiter.api.Test;
+
+/** Unit tests for using the AggregationExtension and ExemplarReservoirFactory internal APIs. */
+public class SdkCustomExemplarReservoirTest {
+  private static final Resource RESOURCE =
+      Resource.create(Attributes.of(stringKey("resource_key"), "resource_value"));
+  private final TestClock testClock = TestClock.create();
+
+  private Meter initialize(InMemoryMetricReader memory, ExemplarReservoirFactory reservoirFactory) {
+    SdkMeterProvider sdkMeterProvider =
+        SdkMeterProvider.builder()
+            .setClock(testClock)
+            .setResource(RESOURCE)
+            .setExemplarFilter(ExemplarFilter.alwaysOn())
+            .registerView(
+                InstrumentSelector.builder().setName("test").build(),
+                View.builder()
+                    .setAggregation(
+                        ((AggregationExtension) Aggregation.sum())
+                            .setExemplarReservoirFactory(reservoirFactory))
```

Comment:
Should exemplar reservoir be a configuration parameter that hangs off an aggregation or a top level view config option? I.e. `View.builder.setExemplarReservoirFactory(...).setAggregation(..)`

Reply:
This is one of the things we need to decide.

- Right now, ExemplarReservoir (defaults) are decided by aggregation.
- If a user wanted to leverage "default" aggregation, but different reservoir by-instrument type or aggregator choice, we're actually lacking an interplay between those two.  E.g. trying to ensure Explicit buckets are the same between exemplars + histograms.

Personally, I think we could go either way here.  I'm happy to extract it out further if we want.

Reply:
Another interesting side affect of having exemplar reservoir be a configurable property of `Aggregation` is that it causes the reservoir be something that MetricReader / MetricExporter have influence over, since MetricReader / MetricExporter can dictate the default `Aggregation` as a function of instrument kind.

Reply:
I actually really like that interaction.  It allows exporters to optimise exemplar choice for the backend they speak to.   Do you think we should encode this in the specification?

Reply:
I think it ought to be in the spec if we're going to allow that type of configuration in the java implementation. 

Another thought: I think we probably should introduce an `AggregationBuilder` instead of allowing `Aggregation` instances to be mutated after instantiation with `setAggregation`.

I propose we add new method `static AggregationBuilder Aggregation#builder()` for accessing an AggregationBuilder, defined as follows:
```
public interface AggregationBuilder {

  // Configure aggregation
  AggregationBuilder setDrop();
  AggregationBuilder setSum();
  AggregationBuilder setLastValue();
  AggregationBuilder setExplicitBucketHistogram();
  AggregationBuilder setExplicitBucketHistogram(List<Double> bucketBoundaries);
  AggregationBuilder setBase2ExponentialHistogram();
  AggregationBuilder setBase2ExponentialHistogram(int maxBuckets, int maxScale);

  // Configure reservoir
  AggregationBuilder setExemplarReservoirFactory(Supplier<ExemplarReservoir<?>> factory);

  Aggregation build();
  
}
```

Usage would be something like:
```
    SdkMeterProvider.builder()
        .registerView(
            InstrumentSelector.builder().build(),
            View.builder()
                .setAggregation(Aggregation.builder()
                    .setSum()
                    .setExemplarReservoirFactory(ExemplarReservoirFactory.noSamples())
                    .build())
                .build());
```